File: src/main/java/org/academy/api/client/gui/animation/AnimationManager.java
package org.academy.api.client.gui.animation;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public final class AnimationManager {
    private static final List<Animator> runningAnimations = new CopyOnWriteArrayList<>();
    private static final List<Animator> pendingAdditions = new CopyOnWriteArrayList<>();

    private AnimationManager() {
    }

    static void startAnimation(Animator animator) {
        pendingAdditions.add(animator);
    }

    static void remove(Animator animator) {
        runningAnimations.remove(animator);
        pendingAdditions.remove(animator);
    }

    public static void onFrameUpdate() {
        var currentTime = System.nanoTime() / 1_000_000;

        if (!pendingAdditions.isEmpty()) {
            for (var anim : pendingAdditions) {
                anim.startTime = currentTime + anim.getStartDelay();
                anim.onStartInternal();
            }
            runningAnimations.addAll(pendingAdditions);
            pendingAdditions.clear();
        }

        if (runningAnimations.isEmpty()) return;

        var toRemove = new ArrayList<Animator>();
        for (var anim : runningAnimations) {
            if (anim.tick(currentTime)) {
                toRemove.add(anim);
            }
        }

        if (!toRemove.isEmpty()) runningAnimations.removeAll(toRemove);
    }
}

File: src/main/kotlin/org/academy/api/client/gui/animation/AnimationManager.kt
package org.academy.api.client.gui.animation

import java.util.concurrent.CopyOnWriteArrayList

object AnimationManager {
    private val runningAnimations: MutableList<Animator> = CopyOnWriteArrayList()
    private val pendingAdditions: MutableList<Animator> = CopyOnWriteArrayList()

    fun startAnimation(animator: Animator) {
        pendingAdditions.add(animator)
    }

    fun remove(animator: Animator) {
        runningAnimations.remove(animator)
        pendingAdditions.remove(animator)
    }

    fun onFrameUpdate() {
        val currentTime = System.nanoTime() / 1000000

        if (!pendingAdditions.isEmpty()) {
            for (anim in pendingAdditions) {
                anim.startTime = currentTime + anim.startDelay
                anim.onStartInternal()
            }
            runningAnimations.addAll(pendingAdditions)
            pendingAdditions.clear()
        }

        if (runningAnimations.isEmpty()) return

        val toRemove = ArrayList<Animator>()
        for (anim in runningAnimations) {
            if (anim.tick(currentTime)) {
                toRemove.add(anim)
            }
        }

        if (!toRemove.isEmpty()) runningAnimations.removeAll(toRemove)
    }
}